### PR TITLE
KT-18074 Suggestion in Intention 'Specify return type explicitly' doesn't support generic type parameter

### PR DESCRIPTION
**Pull request by t-kameyama**
Fix for https://youtrack.jetbrains.com/issue/KT-18074

### CODE REVIEW

**Comment by yole** [comment]
I think it's better to fix the logic so that type parameters can be checked correctly, and not to suppress the check.

**Comment by t-kameyama** [comment]
I fixed to check type parameter.

File: idea/ide-common/src/org/jetbrains/kotlin/idea/util/TypeUtils.kt [hunk 1]
```diff
@@ -97,7 +98,15 @@ fun KotlinType.isResolvableInScope(scope: LexicalScope?, checkTypeParameters: Bo
 
     val descriptor = constructor.declarationDescriptor
     if (descriptor == null || descriptor.name.isSpecial) return false
-    if (!checkTypeParameters && descriptor is TypeParameterDescriptor) return true
+
+    if (descriptor is TypeParameterDescriptor) {
+        if (checkTypeParameters) {
+            val owner = descriptor.containingDeclaration
+            if (owner is FunctionDescriptor && owner.typeParameters.any { it == descriptor }) return true
```
Comment (mglukhikh) [inline idea/ide-common/src/org/jetbrains/kotlin/idea/util/TypeUtils.kt]: any { it == descriptor } is equivalent to just contains(descriptor)

File: idea/ide-common/src/org/jetbrains/kotlin/idea/util/TypeUtils.kt [hunk 2]
```diff
@@ -97,7 +98,15 @@ fun KotlinType.isResolvableInScope(scope: LexicalScope?, checkTypeParameters: Bo
...
+        } else {
```
Comment (mglukhikh) [inline idea/ide-common/src/org/jetbrains/kotlin/idea/util/TypeUtils.kt]: FYI: in Kotlin project if...else is formatted in a different way:
```
if () {
}
else {
}
```

File: idea/ide-common/src/org/jetbrains/kotlin/idea/util/TypeUtils.kt (hunk 1 above)
Comment (t-kameyama) [inline idea/ide-common/src/org/jetbrains/kotlin/idea/util/TypeUtils.kt]: Fixed

File: idea/ide-common/src/org/jetbrains/kotlin/idea/util/TypeUtils.kt (hunk 2 above)
Comment (t-kameyama) [inline idea/ide-common/src/org/jetbrains/kotlin/idea/util/TypeUtils.kt]: Fixed

**Comment by mglukhikh** [comment]
Please pay attention to the fact that function you've changed `KotlinType.isResolvableInScope` is used in different places, not only in the intention you are trying to fix. That's why some logic was broken.

**Comment by t-kameyama** [comment]
I changed to check type parameters in SpecifyTypeExplicitlyIntention.
Please review again.

**Comment by mglukhikh** [comment]
Thank you, I've merged it.

**Pull request closed without merging**